File: apps/api/src/routes/health.ts
import { Router, Request, Response } from 'express';
import { prisma } from '@careerforge/database';
import { getRedisClient } from '../infrastructure/redis/redis.client.js';
import { ApiResponse } from '@careerforge/types';

export const healthRouter: Router = Router();

/**
 * GET /live - Lightweight Liveness Probe
 */
healthRouter.get('/live', (_req: Request, res: Response): void => {
  res.status(200).json({ status: 'alive', uptime: process.uptime() });
});

/**
 * GET /health - Basic Process Health Probe
 */
healthRouter.get('/health', async (req: Request, res: Response): Promise<void> => {
  let dbStatus = 'disconnected';
  try {
    await prisma.$queryRawUnsafe('SELECT 1;');
    dbStatus = 'connected';
  } catch {
    dbStatus = 'disconnected';
  }

  const response: ApiResponse<{
    status: string;
    service: string;
    version: string;
    database: string;
    uptime: number;
    environment: string;
  }> = {
    success: true,
    data: {
      status: 'ok',
      service: 'api',
      version: '1.0.0',
      database: dbStatus,
      uptime: process.uptime(),
      environment: process.env['NODE_ENV'] || 'development',
    },
    meta: {
      correlationId: req.correlationId,
      requestId: req.requestId,
      timestamp: new Date().toISOString(),
    },
  };

  res.status(200).json(response);
});

/**
 * GET /ready - Deep Readiness Probe
 */
healthRouter.get('/ready', async (_req: Request, res: Response): Promise<void> => {
  let isDbReady = false;
  let isRedisReady = false;

  try {
    await prisma.$queryRawUnsafe('SELECT 1;');
    isDbReady = true;
  } catch {
    isDbReady = false;
  }

  try {
    const redis = getRedisClient();
    if (redis && redis.status === 'ready') {
      await redis.ping();
      isRedisReady = true;
    }
  } catch {
    isRedisReady = false;
  }

  const isReady = isDbReady; // DB is critical for readiness; Redis operates with fallback

  const statusCode = isReady ? 200 : 503;
  res.status(statusCode).json({
    status: isReady ? 'ready' : 'not_ready',
    dependencies: {
      database: isDbReady ? 'connected' : 'unreachable',
      redis: isRedisReady ? 'connected' : 'degraded_fallback',
    },
    timestamp: new Date().toISOString(),
  });
});

File: apps/ai-service/app/api/routes/health.py
from datetime import datetime, timezone
from fastapi import APIRouter, Request, Response, status
from ...core.config import settings
from ...schemas.health import HealthResponse, HealthResponseData, HealthMeta
from ...services.vector_store import FAISSVectorStore

router = APIRouter()

@router.get("/live")
async def get_liveness():
    """Lightweight process liveness probe."""
    return {"status": "alive", "service": "ai-service"}

@router.get("/health", response_model=HealthResponse)
async def get_health(request: Request):
    """Health check endpoint returning service status, runtime environment, and active LLM provider."""
    correlation_id = getattr(request.state, "correlation_id", None)
    request_id = getattr(request.state, "request_id", None)
    
    return HealthResponse(
        success=True,
        data=HealthResponseData(
            status="ok",
            service="ai-service",
            version="1.0.0",
            environment=settings.ENVIRONMENT,
            provider=settings.LLM_PROVIDER
        ),
        meta=HealthMeta(
            correlationId=correlation_id,
            requestId=request_id,
            timestamp=datetime.now(timezone.utc).isoformat()
        )
    )

@router.get("/ready")
async def get_readiness(response: Response):
    """Deep readiness probe verifying FAISS vector store and embedding provider initialization."""
    is_vector_ready = False
    details = {}

    try:
        store = FAISSVectorStore.get_instance()
        stats = store.get_stats()
        is_vector_ready = store.index is not None
        details["vector_store"] = {
            "status": "ready" if is_vector_ready else "not_initialized",
            "total_vectors": stats.get("total_vectors", 0),
            "dimension": stats.get("embedding_dimension", 384),
        }
    except Exception as e:
        details["vector_store"] = {"status": "error", "message": str(e)}

    is_ready = is_vector_ready

    if not is_ready:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE

    return {
        "status": "ready" if is_ready else "not_ready",
        "service": "ai-service",
        "provider": settings.LLM_PROVIDER,
        "embedding_provider": settings.EMBEDDING_PROVIDER,
        "dependencies": details,
        "timestamp": datetime.now(timezone.utc).isoformat()
    }
